Qprof step 10 (work in progress)

jslaunwhite-microfocus opened the pull request:
## VER-90918
JIRA https://jira.verticacorp.com/jira/browse/VER-90918

* Add the query events report to qprof.py
* Add some tests for qprof.py, including some small test fixtures

TODO
* Update the docstring for query events - it is missing pieces
* Add a test that actually tests query events ;)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vertica/VerticaPy?pullRequest=1038) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vertica/VerticaPy?pullRequest=1038) before we can accept your contribution.<br/><hr/>**Jason Slaunwhite** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vertica/VerticaPy?pullRequest=1038) it.</sub>

vikash018 commented:
@jslaunwhite-microfocus - Please add Copyright statement at the beginning of each file. you can refer to any existing file.

oualib commented:
@jslaunwhite-microfocus please use the latest version of black to format your code.
@vikash018 please look closely at the tests to see if it follows your logic.
I will now look at the main code.

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -1662,6 +1662,47 @@ def get_qplan_profile(
+        """
+        Returns a vDataFrame that contains a table listing query events
+
+        Parameters
```
when we don't have parameters, we skip the section.

jslaunwhite-microfocus replied:
Skipped this section

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -1662,6 +1662,47 @@ def get_qplan_profile(
+    # Step 10: Query events
+    def get_query_events(self) -> vDataFrame:
+        """
+        Returns a vDataFrame that contains a table listing query events
```
You can use ```:py:class`vDataFrame` ``` to create a reference.

Also make it on two lines as it is recommended to have less than 70 characters by line (maybe I am wrong, but it should be close to this number).

```
Returns a :py:class`vDataFrame` that 
contains a table listing query events
```

jslaunwhite-microfocus replied:
Updated to use :py:class`vDataFrame`

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -1662,6 +1662,47 @@ def get_qplan_profile(
+        -------
+        A vDataframe that contains a table listing query events
+
+        Examples
```
Yes, and please, add a more detailed example in the main class example above.

jslaunwhite-microfocus replied:
I have added a detailed example the query events function that I added.

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -1662,6 +1662,47 @@ def get_qplan_profile(
+        """
+        query = self._replace_schema_in_query(query)
+        vdf = vDataFrame(query)
+        return vdf
```
Good!

oualib commented on verticapy/tests_new/performance/vertica/conftest.py:
```diff
@@ -0,0 +1,72 @@
+import pytest
```
@vikash018 can you please look at it?

oualib commented on verticapy/tests_new/performance/vertica/test_conn_simple.py:
```diff
@@ -0,0 +1,81 @@
+import pytest
```
@vikash018 can you please look at it?

oualib reviewed (commented):
Small changes to make on the doc.
@vikash018 please review the tests.

jslaunwhite-microfocus commented:
> @jslaunwhite-microfocus - Please add Copyright statement at the beginning of each file. you can refer to any existing file.

Fixed with 6be2d6b9e481e99cf776e779d299234c227c1382

jslaunwhite-microfocus commented:
> @jslaunwhite-microfocus please use the latest version of black to format your code. @vikash018 please look closely at the tests to see if it follows your logic. I will now look at the main code.

Fixed with e76f71c56d7526719ad6ff1d38c0d9aa95afd572

vikash018 commented on verticapy/tests_new/performance/vertica/test_conn_simple.py:
```diff
@@ -0,0 +1,105 @@
+    assert fragment.lower() in sql.lower()
+
+
+def test_connect_perf(vp_connect_get_name, raw_client_cursor):
```
@jslaunwhite-microfocus - it would be helpful to add pytest.mark.parametrize as there are multiple assert.

jslaunwhite-microfocus replied:
This has been removed. 

As discussed, parameterized isn't exactly what we're looking for here. Instead, the sequence of asserts confirms that the object behaves as expected.

vikash018 commented on verticapy/tests_new/performance/vertica/test_conn_simple.py:
```diff
@@ -0,0 +1,105 @@
+    check_request(qp, "SELECT COUNT(*)")
+
+
+def setup_dummy_table(raw_client_cursor, table_name):
```
@jslaunwhite-microfocus - we run test in parallel. It would be better to create unique schema to avoid conflicts. You can use existing fixture load_test_schema that can server this purpose.

jslaunwhite-microfocus replied:
The statement was removed.

vikash018 reviewed (commented):
Overall unit test looks good. Have added few comments, pls do needful.

oualib commented on verticapy/tests_new/performance/vertica/conftest.py:
```diff
@@ -0,0 +1,98 @@
+            conn.close()
+
+
+def get_connection_info(verbose=True):
```
@vikash018 do we need to recreate connection functions each time?

vikash018 replied:
@jslaunwhite-microfocus - can you try using current cursor? This is how currently we are doing if we need to run sql statement. 
from verticapy.connection import current_cursor
current_cursor().execute("select * from table" )

jslaunwhite-microfocus replied:
I removed this and replaced it with asserts

`assert len(vp.available_connections()) > 0`

jslaunwhite-microfocus replied:
I removed the connection setup and started using the connection setup recommended by @afard . I am able to get the desired behavior for my tests. It took some fussing with my environment to get things right, but now it is all set. You can see that the conftest.py file has been removed.

oualib reviewed (commented):
@vikash018 please double check

vikash018 commented on verticapy/tests_new/performance/vertica/conftest.py:
```diff
@@ -0,0 +1,98 @@
+    logging.info("Connection complete")
+    return connection_name
+
+
```
@jslaunwhite-microfocus - same goes here as line# 53.

vikash018 commented on verticapy/tests_new/performance/vertica/test_qprof_simple.py:
```diff
@@ -0,0 +1,100 @@
+    check_request(qp, "avg(number) AS avg_number")
+
+
+def setup_dummy_table_run_query() -> vDataFrame:
```
we have data load fixture define in conftest file present at VerticaPy/verticapy/tests_new/conftest.py . Please use it. if you want to define any new data and can be shared by other test. Please defined under this conftest file.

jslaunwhite-microfocus replied:
schema_loader added with commit 42001a92

vikash018 replied:
setup_dummy_table_run_query function is not needed. You can directly use amazon_vd fixture defined in VerticaPy/verticapy/tests_new/conftest.py.

jslaunwhite-microfocus commented:
@afard @vikash018 @mail4umar - I have addressed all of the comment. The checks are queued and waiting to run. Can you please take another look and let me know if there is more that I should do prior to merge?

oualib commented:
@jslaunwhite-microfocus I've made the modifications (test + docstrings - now it looks good except something). For some reasons your test is failing because it did not connect to the server (no available connection).

BESIDES: I prefer one class test for all QUERYPROFILER:

@vikash018 can you please set up a call with @jslaunwhite-microfocus to see how you can together correct the test file. Be sure to have something consistent. It is preferable to change the overall structure to have 1 class for the QueryProfiler objects tests and to do ALL the tests under that class.

jslaunwhite-microfocus commented:
> @jslaunwhite-microfocus I've made the modifications (test + docstrings - now it looks good except something). For some reasons your test is failing because it did not connect to the server (no available connection).

Yes, I see this failure, and it surprises me. I'll investigate again with @afard  and @vikash018 .

> 
> BESIDES: I prefer one class test for all QUERYPROFILER:

> @vikash018 can you please set up a call with @jslaunwhite-microfocus to see how you can together correct the test file. Be sure to have something consistent. It is preferable to change the overall structure to have 1 class for the QueryProfiler objects tests and to do ALL the tests under that class.

Can you tell me about what you mean by "one class test"? I'm happy to do whatever the standard is. This PR adds a small test for part of the query profiler. I was planning to add more small tests as we go along. I was planning to add each test as a function, either in `test_qprof_simple.py` file or in another file in this directory. That's how I've seen other tests handled before, and I thought that was the structure of `tests_new/`

jslaunwhite-microfocus commented:
> 
> Can you tell me about what you mean by "one class test"? I'm happy to do whatever the standard is. This PR adds a small test for part of the query profiler. I was planning to add more small tests as we go along. I was planning to add each test as a function, either in `test_qprof_simple.py` file or in another file in this directory. That's how I've seen other tests handled before, and I thought that was the structure of `tests_new/`

Nevermind, I see the changes that @oualib pushed. The comment makes sense with that example. Sure, I can organize things that way

oualib commented:
> > Can you tell me about what you mean by "one class test"? I'm happy to do whatever the standard is. This PR adds a small test for part of the query profiler. I was planning to add more small tests as we go along. I was planning to add each test as a function, either in `test_qprof_simple.py` file or in another file in this directory. That's how I've seen other tests handled before, and I thought that was the structure of `tests_new/`
> 
> Nevermind, I see the changes that @oualib pushed. The comment makes sense with that example. Sure, I can organize things that way

I am not the Unit Test master but @vikash018 can confirm.

Also @vikash018 can you please write a RST file with an example on how you expect test files to be organised?

jslaunwhite-microfocus commented:
@oualib -- test 3.11 was failing before, and it is passing now. I hope the other tests pass soon. What will be the next steps to move this forward?

jslaunwhite-microfocus commented:
> @oualib -- test 3.11 was failing before, and it is passing now. I hope the other tests pass soon. What will be the next steps to move this forward?

@oualib @afard and @mail4umar and @vikash018  - all checks passed, I think this is ready to merge. Just need a review.

oualib reviewed (approved):
Approved. Looks good we can merge.

jslaunwhite-microfocus commented:
Thank you! I am very glad to have this merged.
